File: python/hyperbug/device.py
"""Base classes for hyperbug device plugins.

hyperbug's Rust side loads a device plugin file, instantiates a named
class from it, and calls `read`/`write` by plain attribute/method access
(see src/pydevice.rs) — subclassing these base classes is not required for
that to work, but gets you the documented contract, type hints, and a
clear place to look up what's expected instead of it living only in a
Rust doc comment.
"""

from abc import ABC, abstractmethod

#: The newest plugin ABI version this hyperbug build implements. Bumped
#: only when a *breaking* change is made to the `Device`/`PciDevice`
#: contract itself (a required method's signature, the meaning of an
#: existing attribute, ...) — not for additions like `tick()`, which are
#: purely opt-in and never break an existing plugin that doesn't define
#: them. `Device`/`PciDevice` subclasses pick this up automatically as
#: `self.hyperbug_api_version`; hyperbug's Rust side checks it once at
#: load (see `src/pydevice.rs`/`src/pydevice_proc.rs`) against the
#: inclusive range `[HYPERBUG_API_MIN_SUPPORTED, HYPERBUG_API_VERSION]` —
#: declaring something newer means the plugin needs a newer hyperbug;
#: declaring something older than `HYPERBUG_API_MIN_SUPPORTED` means
#: hyperbug has since dropped support for that contract. Either fails
#: loudly at load, saying which, instead of silently running the plugin
#: against a shape it never agreed to. A plugin that doesn't subclass
#: `Device` at all (duck-typing is explicitly supported — see the module
#: docstring) has no version to check, and loads exactly as before this
#: existed.
HYPERBUG_API_VERSION = 1

#: The oldest `hyperbug_api_version` a plugin may declare and still be
#: accepted — see `HYPERBUG_API_VERSION` above. Equal to it today (only
#: one contract has ever existed); bumped only alongside
#: `HYPERBUG_API_VERSION` when an old, numbered contract can no longer be
#: supported.
HYPERBUG_API_MIN_SUPPORTED = 1


class Device(ABC):
    """A memory-mapped (or I/O-port-mapped) device.

    Register with hyperbug via ``--device <path>:<ClassName>:<base>:<size>``
    for a fixed-address MMIO device, or as part of a `PciDevice` for one
    whose address comes from PCI enumeration instead.

    Every instance also gets a ``self.hyperbug`` attribute set right after
    construction (not available inside ``__init__`` — defer anything that
    needs it to the first `read`/`write` call instead), giving access to:

    - ``self.hyperbug.read_mem(addr, size) -> bytes`` /
      ``self.hyperbug.write_mem(addr, data)`` — real DMA against guest
      physical memory, bounds-checked (raises `OSError` on an
      out-of-range address, since a guest-supplied address is untrusted
      input). Lets a device act like real hardware would: read a
      descriptor the guest posted, write a response into a
      guest-supplied buffer, instead of only ever exposing its own
      internal register state.

      **Also raises `OSError`** if the operator launched this plugin with
      a declared DMA-confinement range (the optional trailing
      ``:<dma_base>:<dma_size>`` on ``--device``/``--pci-device`` and
      their ``-sandboxed`` variants) and `addr` falls outside it — a
      host-side vIOMMU-shaped control, invisible to the guest, that
      bounds what this specific plugin instance may touch regardless of
      what its own code does. Not something a plugin can query or opt out
      of from inside itself: whether it's applied, and to what range, is
      entirely the launching operator's decision. Most plugins (including
      every one shipped in `devices/`) run with no such restriction and
      never see this — it exists for confining a numerous or third-party
      plugin an operator doesn't fully trust with all of guest RAM.
    - ``self.hyperbug.raise_irq()`` — requests this device's configured
      interrupt (see `PciDevice.interrupt_line` below) *spontaneously*,
      any time, not only synchronously from inside `write()` the way a
      truthy return does. This is a **separate mechanism** from `write`'s
      return value below — a device that raises its interrupt via
      `raise_irq()` should not also return `True`, and vice versa; mixing
      them up is an easy mistake (see `devices/dma_demo.py` for a worked
      example using `raise_irq()`, and `devices/scratch.py` for the
      plain-return-value style).
    """

    #: See `HYPERBUG_API_VERSION` above — checked once at load, not meant
    #: to be overridden by a plugin.
    hyperbug_api_version: int = HYPERBUG_API_VERSION

    @abstractmethod
    def read(self, offset: int, size: int) -> bytes:
        """Return exactly `size` bytes read from `offset` (relative to
        this device's own base address, not the raw guest address)."""
        raise NotImplementedError

    @abstractmethod
    def write(self, offset: int, data: bytes) -> bool | None:
        """Handle `data` written at `offset`.

        Return `True` to request that this device's configured interrupt
        (if it has one — only meaningful for a `PciDevice` registered via
        `--pci-device`) be raised right now, synchronously. Returning
        `None` (i.e. not returning anything) is the same as `False`. See
        `self.hyperbug.raise_irq()` above for the spontaneous alternative.
        """
        raise NotImplementedError

    # `tick(self) -> None` is *not* defined here, deliberately: hyperbug
    # checks once at load whether a plugin defines one **at all** (see
    # `check_api_version`'s sibling check in `src/pydevice.rs`/
    # `_sandbox_runner.py`) and only ever calls a plugin that does — giving
    # `Device` itself a no-op `tick` would make every subclass "define
    # one" whether it overrides it or not, defeating that. If you want
    # one, just add:
    #
    #     def tick(self) -> None:
    #         """Called once per host vCPU-loop iteration, independent of
    #         any guest register access. Use it for logic that shouldn't
    #         wait for the guest to touch a register first: a simulated
    #         timer, or an asynchronous transaction (post a request, then
    #         some time later — not necessarily in response to any guest
    #         write — call `self.hyperbug.raise_irq()` once it completes).
    #         See `devices/doorbell_demo.py` for a worked example of the
    #         latter."""

    # `reset(self) -> None` is *not* defined here either, same reasoning
    # and same opt-in check (`HAS_RESET`) as `tick` above. The difference
    # from `tick`: nothing in hyperbug calls this automatically — there's
    # no in-guest reset trigger modeled today (no PCI function-level
    # reset, no guest-driver-initiated reset). The only real caller is the
    # live control socket's `reset_device <mmio|pci> <selector>` command
    # (see `hyperbug.vm.ControlConnection.reset_device`) — an operator
    # explicitly asking this specific device to reinitialize. If you want
    # one, just add:
    #
    #     def reset(self) -> None:
    #         """Reinitialize this device's internal state, analogous to
    #         a real hardware reset line. Called only when an operator
    #         explicitly asks for it via the control socket — never
    #         automatically."""


class PciDevice(Device):
    """A `Device` that also declares a PCI identity, for
    ``--pci-device <path>:<ClassName>:<device_number>``.

    hyperbug's own code has no opinion on what any specific PCI device is
    — vendor ID, device ID, and everything else about a real device's
    identity belongs here, in your plugin, not in hyperbug's Rust core.

    Set these as class or instance attributes. hyperbug reads them
    **once**, when the device is loaded, and serves PCI config space from
    that snapshot afterwards — the same way a real device's config-space
    identity bytes are fixed in silicon. Changing one of these attributes
    later, at runtime, has no effect (and reading them per config-space
    access would mean taking the GIL for every dword the guest touches
    while enumerating the bus).
    """

    #: 16-bit PCI vendor ID.
    vendor_id: int
    #: 16-bit PCI device ID.
    device_id: int
    #: Packed 24-bit (base class << 16 | subclass << 8 | prog-if) PCI
    #: class code, e.g. 0x020000 for an Ethernet controller.
    class_code: int
    #: BAR sizes in bytes, BAR0 first; up to 6 entries, 0 (or omitted)
    #: meaning "this BAR isn't implemented".
    bar_sizes: list[int]
    #: Optional: BAR indices (0-5) that are I/O-space rather than
    #: memory-space. Omit entirely (the default) if every BAR is
    #: memory-space, which is what real hardware almost always is outside
    #: of legacy devices like virtio's own BAR0.
    io_bars: list[int]
    #: Optional: the legacy ISA IRQ (1-15) this device's interrupt is
    #: wired to. Without one, both `write()`'s truthy-return path and
    #: `self.hyperbug.raise_irq()` have nothing to pulse (0, the default,
    #: means "no interrupt" throughout hyperbug — see `pci.rs`). Must
    #: actually be routed: a `--pci-device` slot outside the ranges
    #: `acpi/dsdt.asl`'s `_PRT` already covers (disks 4-11, net 16, rng
    #: 17) has no ACPI routing entry for whatever IRQ you pick here, so it
    #: won't deliver under ACPI without adding one there too — it will
    #: still work with ACPI disabled, which trusts this value directly.
    interrupt_line: int
    #: Optional: set `True` to give this device a real PCI MSI capability
    #: (single message, 32-bit address, no per-vector masking — the
    #: simplest real form; not MSI-X). If a guest driver enables it,
    #: `write()`'s truthy return and `self.hyperbug.raise_irq()` both
    #: deliver via a real `KVM_SIGNAL_MSI` instead of pulsing
    #: `interrupt_line` — transparent to your device code either way, you
    #: still just return `True` or call `raise_irq()`. See
    #: `devices/dma_demo.py` for a worked example. **Not useful for
    #: emulating virtio**: hyperbug's own virtio devices deliberately
    #: never set this — checked directly against the real Linux virtio
    #: driver source, which only ever requests MSI-X or falls back to
    #: plain INTx, never single-vector MSI, so a virtio-shaped device
    #: would never actually have this used by a real driver. It's for a
    #: *custom* device model with its own (real or hypothetical) driver
    #: that genuinely requests plain MSI.
    msi_capable: bool


class I2cDevice(ABC):
    """A target device on hyperbug's single virtual I2C bus.

    Register with ``--i2c-device <path>:<ClassName>:<addr>`` (a 7-bit
    address, 0x00-0x7f). The guest reaches this through a real
    ``virtio-i2c`` adapter (``VIRTIO_ID_I2C_ADAPTER``) that its own
    unmodified ``i2c-virtio`` kernel driver binds to -- no custom guest
    driver needed, and no fixed register-offset addressing the way
    `Device` has: I2C messages are just address + direction + a byte
    buffer, matching the guest's own ``i2c_msg`` abstraction.

    Unlike `Device`, an `I2cDevice` gets **no** ``self.hyperbug`` DMA
    context -- it never sees guest memory directly, only the bytes of
    whichever message was addressed to it. Also unlike `Device`, there's
    no PCI identity to declare here: the *adapter* (one per hyperbug
    process, created automatically the moment any `I2cDevice` is
    attached) is what the guest's PCI core sees; individual targets are
    invisible to PCI enumeration, exactly like real I2C peripherals.

    Real I2C transactions frequently chain a write (setting an internal
    "register pointer") immediately followed by a read (returning data
    from that pointer) with no bus release in between -- the standard way
    almost every real sensor works. hyperbug delivers `i2c_write`/
    `i2c_read` calls to the *same* long-lived instance, strictly in the
    order the guest issued them, which is enough for a stateful device to
    implement that idiom correctly with nothing more than an instance
    attribute (see `devices/i2c_temp_sensor.py` for a worked example) --
    there's no separate "transaction" object or start/stop callback to
    manage.
    """

    #: See `HYPERBUG_API_VERSION` above -- checked once at load, same as
    #: `Device`.
    hyperbug_api_version: int = HYPERBUG_API_VERSION

    @abstractmethod
    def i2c_write(self, data: bytes) -> None:
        """Handle a write-direction message: `data` is exactly what the
        guest wrote in one `i2c_msg` (may be empty -- a real I2C master
        can issue a zero-length write purely to probe whether a device
        answers this address at all; that case never reaches this
        method, since hyperbug ACKs/NAKs a zero-length probe from
        whether a device is registered at all, with no `i2c_write`/
        `i2c_read` call either way)."""
        raise NotImplementedError

    @abstractmethod
    def i2c_read(self, length: int) -> bytes:
        """Return up to `length` bytes for a read-direction message.
        Returning fewer than `length` bytes is fine (only the bytes you
        return are transferred); returning more is truncated."""
        raise NotImplementedError


class GpioBank(ABC):
    """One virtual GPIO chip on hyperbug's guest-facing GPIO surface.

    Register with ``--gpio-device <path>:<ClassName>`` -- unlike
    `I2cDevice` (many targets sharing one bus), each `GpioBank` gets its
    own independent PCI slot and its own real ``virtio-gpio`` adapter
    (``VIRTIO_ID_GPIO``), matching real BMC hardware's typically several
    separate GPIO controllers rather than one shared bus.

    Every instance also gets a ``self.hyperbug`` attribute (set right
    after construction, same timing as `Device`'s) with exactly one
    method: ``self.hyperbug.raise_irq(line)``, which marks `line` as
    having spontaneously changed state. It's delivered to the guest as a
    real interrupt only if that line currently has one armed (the guest
    posted a buffer for it) and enabled (`IRQ_TYPE` isn't `NONE`) --
    otherwise dropped, matching a real masked/disabled interrupt. This is
    real eventfd-backed and thread-safe to call from anywhere, including
    your own background Python thread (`threading.Timer`, a polling
    loop) -- there's no per-iteration `tick()` hook the way `Device` has,
    since this already covers the same need with no extra polling path.
    """

    #: See `HYPERBUG_API_VERSION` above.
    hyperbug_api_version: int = HYPERBUG_API_VERSION

    #: Number of lines this bank exposes. Fixed for the life of the bank
    #: -- read once when the plugin is loaded.
    ngpio: int

    #: Optional per-line names, exposed to the guest via
    #: `GPIO_V2_GET_LINEINFO_IOCTL`-shaped introspection. Either omit
    #: entirely (no names offered at all -- the real driver never even
    #: asks) or supply exactly `ngpio` entries.
    names: list[str] = []

    @abstractmethod
    def get_direction(self, line: int) -> int:
        """Return one of `hyperbug.gpio`'s `DIRECTION_NONE`/`DIRECTION_
        OUT`/`DIRECTION_IN` for `line`."""
        raise NotImplementedError

    @abstractmethod
    def set_direction(self, line: int, direction: int) -> None:
        """`direction` is one of `DIRECTION_NONE`/`DIRECTION_OUT`/
        `DIRECTION_IN` -- already validated before this is called."""
        raise NotImplementedError

    @abstractmethod
    def get_value(self, line: int) -> int:
        """Return 0 or 1 for `line`'s current value."""
        raise NotImplementedError

    @abstractmethod
    def set_value(self, line: int, value: int) -> None:
        """`value` is 0 or 1 -- already masked before this is called."""
        raise NotImplementedError

File: src/device.rs
//! The scriptable device ABI: any virtual device (native Rust or a Python
//! plugin) implements `Device`, and a `Bus` dispatches guest accesses to
//! whichever registered device owns that address/port range. `MmioBus` and
//! `IoBus` are the same dispatch logic — one keyed on guest-physical
//! addresses, the other on 16-bit I/O ports (legacy virtio's BAR0 is
//! I/O-space, not memory) — so they share one implementation.
//!
//! Devices are held as `Arc<Mutex<dyn Device>>` rather than owned outright,
//! since a PCI device's BAR-mapped registers need to be reachable both from
//! a `Bus` (for the actual reads/writes) and from `PciBus` (which moves the
//! mapping around as the guest sizes and programs the BAR).
//!
//! `Arc<Mutex<...>>`, not `Rc<RefCell<...>>` (DEBTS.md item 11, SMP):
//! multiple vCPU threads can trigger accesses to the same device
//! concurrently once there's more than one vCPU — the lock is what makes
//! that safe, and `Arc`/`Mutex` (rather than `Rc`/`RefCell`) is what makes
//! it *possible* to share a device across threads at all (`Rc` is never
//! `Send`, regardless of what guards it).

use std::sync::{Arc, Mutex};

/// The `Device`/`PciDevice` plugin ABI version this build implements — kept
/// in lockstep with `python/hyperbug/device.py`'s `HYPERBUG_API_VERSION`,
/// the single source of truth for what the number means and when to bump
/// it. Checked once at plugin load time (`pydevice.rs`/`pydevice_proc.rs`)
/// against a plugin's own `hyperbug_api_version` attribute, so a plugin
/// written against a since-changed contract fails loudly at load instead
/// of silently misbehaving against a `Device`/`PciDevice` shape that's
/// moved out from under it.
pub const PLUGIN_API_VERSION: u32 = 1;

/// Whether `[addr, addr + len)` lies entirely within a device's declared
/// DMA-confinement range (`config::DeviceSpec`/`PciDeviceSpec`'s optional
/// `dma_range`, a `(base, size)` pair) — the host-side enforcement half of
/// a real, if narrowly-scoped, vIOMMU-shaped control: any Python device
/// plugin's `self.hyperbug.read_mem`/`write_mem` can otherwise touch all
/// of guest RAM, which is fine for a trusted first-party plugin but not
/// once plugins are numerous or third-party (a compromised/malicious
/// plugin could read or corrupt memory belonging to anything else in the
/// guest). `range` is `None` for a plugin launched with no confinement —
/// unrestricted, the only behavior that existed before this — in which
/// case this always returns `true`.
///
/// This is *not* a guest-visible IOMMU: the guest's own OS/drivers see
/// nothing about it, and a device's own descriptor-supplied buffer
/// addresses (what virtio devices are handed by the guest) still need
/// broad reach by design — a real per-guest-programmed IOMMU (DMAR/IVRS
/// tables, an emulated translation-table walker) is a much larger, and
/// still entirely undone, feature. This is deliberately the smaller,
/// tractable piece: an *operator*-declared bound (on the CLI, not trusted
/// from the plugin's own Python attributes — the plugin is exactly the
/// untrusted party here, so it can't be allowed to declare its own
/// confinement) on what a specific plugin instance may touch.
#[inline]
pub fn dma_range_allows(range: Option<(u64, u64)>, addr: u64, len: u64) -> bool {
    let Some((base, size)) = range else { return true };
    let Some(end) = addr.checked_add(len) else { return false };
    let Some(range_end) = base.checked_add(size) else { return false };
    addr >= base && end <= range_end
}

/// A device addressed by an offset-relative-to-its-own-base range. `offset`
/// is relative to that base, not the raw guest-physical address or port.
/// `Send`: required for `Arc<Mutex<dyn Device>>` itself to be shareable
/// across the per-vCPU threads that call into it.
pub trait Device: Send {
    fn read(&mut self, offset: u64, data: &mut [u8]);
    /// Returns `true` if this write should raise the device's configured
    /// interrupt right now (e.g. a virtio device just advanced its used
    /// ring, or a Python device's `write()` returned a truthy value). A
    /// device with no configured IRQ (see `Bus::register`) has this
    /// silently ignored rather than needing to know that itself.
    fn write(&mut self, offset: u64, data: &[u8]) -> bool;

    /// Checked once per vCPU-loop iteration for a spontaneous interrupt
    /// this device wants to raise outside of `write()` (a Python plugin's
    /// `hyperbug.raise_irq()`, in-process or sandboxed). Clears whatever
    /// pending state it reports either way. Most devices never raise one
    /// this way, hence the default; `PyDevice`/`SandboxedPyDevice` are the
    /// only overrides today.
    fn take_pending_irq(&self) -> bool {
        false
    }

    /// Called once per vCPU-loop iteration, before `take_pending_irq`, for
    /// logic that needs to run independent of any guest register access —
    /// e.g. a simulated timer, or an async transaction (a doorbell-style
    /// device bridging to something that completes on its own schedule)
    /// whose completion isn't driven by the guest touching a register at
    /// all (DEBTS.md item 32). Most devices have nothing to do here, hence
    /// the default; a `PyDevice`/`SandboxedPyDevice` whose plugin class
    /// defines `tick()` are the only overrides today.
    fn tick(&mut self) {}
}

/// A shared, lockable handle to a device — the only form a `Bus` (or
/// `PciBus`) ever holds one in.
pub type SharedDevice = Arc<Mutex<dyn Device>>;

struct Region {
    base: u64,
    /// The last address this region covers, **inclusive**. Deliberately
    /// not "one past the end": a device mapped against the very top of
    /// the address space has no representable exclusive end, and
    /// saturating one at `u64::MAX` silently drops the final byte from
    /// the region (which is exactly what the overflow-region test caught).
    /// An inclusive end can express every legal region without wrapping,
    /// and — unlike a saturating exclusive end — stays correct even when
    /// checking two regions that both reach `u64::MAX` for overlap.
    last: u64,
    device: SharedDevice,
    irq: Option<u32>,
}

impl Region {
    /// `addr` is covered iff it falls in `[base, last]` inclusive — not
    /// `addr < base + size`, which is exactly the saturating-exclusive-end
    /// comparison that excludes `u64::MAX` itself from a region that
    /// legitimately reaches it (a BAR placed at the very top of the
    /// guest's address space) — caught by a real test, not found by
    /// inspection.
    #[inline]
    fn contains(&self, addr: u64) -> bool {
        addr >= self.base && addr <= self.last
    }
}

#[derive(Default)]
pub struct Bus {
    regions: Vec<Region>,
}

pub type MmioBus = Bus;
pub type IoBus = Bus;

impl Bus {
    pub fn new() -> Self {
        Self::default()
    }

    /// Registers `device` to handle addresses/ports `[base, base + size)`.
    /// If `device` already has a mapping (e.g. a PCI BAR being
    /// reprogrammed), that old mapping is replaced rather than left stale.
    /// `irq`, if given, is the GSI to report from `write()` when the
    /// device requests an interrupt.
    pub fn register(&mut self, base: u64, size: u64, device: SharedDevice, irq: Option<u32>) {
        self.unregister(&device);
        if size == 0 {
            return; // a zero-sized region maps nothing
        }
        // Saturating, not wrapping: a guest is free to program a BAR near
        // the top of the address space, and overflowing here would
        // otherwise turn the overlap test below into a no-op and make
        // `contains` reject every address. Saturation only ever *shrinks*
        // an already-impossible region (one running past `u64::MAX`), so
        // no legal mapping loses an address to it.
        let last = base.saturating_add(size - 1);
        // Two inclusive ranges overlap iff each starts at or before the
        // other's end.
        if let Some(clash) = self.regions.iter().find(|r| base <= r.last && r.base <= last) {
            // A misbehaving guest (or plugin) shouldn't take the whole VMM
            // down over this: refuse the new mapping and leave whatever
            // was already there working, rather than panicking.
            eprintln!(
                "[hyperbug] refusing to map [{base:#x}, {last:#x}]: overlaps existing mapping \
                 [{:#x}, {:#x}]",
                clash.base, clash.last
            );
            return;
        }
        self.regions.push(Region { base, last, device, irq });
    }

    /// Drops `device`'s mapping, if it has one. Used when a PCI BAR is
    /// reprogrammed back to "unassigned" (base 0): leaving the old mapping
    /// live would keep answering accesses at an address the guest has
    /// explicitly reclaimed.
    pub fn unregister(&mut self, device: &SharedDevice) {
        self.regions.retain(|r| !Arc::ptr_eq(&r.device, device));
    }

    #[inline]
    fn find(&mut self, addr: u64) -> Option<(&mut Region, u64)> {
        let region = self.regions.iter_mut().find(|r| r.contains(addr))?;
        let offset = addr - region.base;
        Some((region, offset))
    }

    /// Looks up the device (if any) covering `addr` and returns a cloned
    /// handle — plus its offset and configured IRQ — *without* locking or
    /// calling into the device itself. Exists so a caller holding a
    /// coarser lock around this `Bus` (`SharedState`'s, in practice) can
    /// release it before actually invoking the device: a device's own
    /// `read`/`write` can be slow (a Python plugin runs arbitrary guest
    /// code under the GIL; a sandboxed plugin is a subprocess round trip,
    /// up to its own 200ms timeout) and must not block every *other*
    /// device sharing that outer lock for the whole duration — see
    /// `vcpu::Runner::dispatch_write`/`dispatch_read`, the only callers.
    #[inline]
    pub fn find_device(&mut self, addr: u64) -> Option<(SharedDevice, u64, Option<u32>)> {
        let (region, offset) = self.find(addr)?;
        Some((region.device.clone(), offset, region.irq))
    }

    /// Returns true if some registered device handled the access. A
    /// combined find-then-call convenience; production code uses
    /// `find_device` instead so it can release its own lock around this
    /// `Bus` before calling into a device (see `find_device`'s doc
    /// comment) — this stays test-only rather than as unused production
    /// API.
    #[cfg(test)]
    pub fn read(&mut self, addr: u64, data: &mut [u8]) -> bool {
        match self.find(addr) {
            Some((region, offset)) => {
                region.device.lock().unwrap().read(offset, data);
                true
            }
            None => false,
        }
    }

    /// `None` if no registered device covers `addr` at all. `Some(irq)`
    /// (where `irq` is `None` if the write didn't request one, or none is
    /// configured for that region) if some device did handle it. Test-only
    /// — see `read`'s doc comment.
    #[cfg(test)]
    pub fn write(&mut self, addr: u64, data: &[u8]) -> Option<Option<u32>> {
        let (region, offset) = self.find(addr)?;
        let wants_irq = region.device.lock().unwrap().write(offset, data);
        Some(wants_irq.then_some(region.irq).flatten())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dma_range_allows_only_ranges_fully_inside_the_declared_bound() {
        assert!(dma_range_allows(None, 0, u64::MAX), "no declared range means unrestricted");

        let range = Some((0x1000, 0x1000)); // [0x1000, 0x2000)
        assert!(dma_range_allows(range, 0x1000, 0x1000), "exactly the whole range");
        assert!(dma_range_allows(range, 0x1500, 0x100), "a sub-range");
        assert!(!dma_range_allows(range, 0x0f00, 0x100), "starts before the range");
        assert!(!dma_range_allows(range, 0x1f00, 0x200), "ends past the range");
        assert!(!dma_range_allows(range, 0, 0x2000), "spans the whole range and beyond");
        assert!(!dma_range_allows(range, u64::MAX - 15, 32), "an overflowing request");
    }

    struct Dummy;

    impl Device for Dummy {
        fn read(&mut self, _offset: u64, data: &mut [u8]) {
            data.fill(0xa5);
        }
        fn write(&mut self, _offset: u64, _data: &[u8]) -> bool {
            true
        }
    }

    fn dummy() -> SharedDevice {
        Arc::new(Mutex::new(Dummy))
    }

    #[test]
    fn a_region_whose_end_would_overflow_still_dispatches_correctly() {
        // A guest is free to program a BAR near the top of the address
        // space; `base + size` wrapping used to make `contains` reject
        // every address and turn the overlap check into a silent no-op.
        let mut bus = Bus::new();
        bus.register(u64::MAX - 15, 64, dummy(), None);

        let mut data = [0u8; 4];
        assert!(bus.read(u64::MAX - 15, &mut data), "the base address must dispatch");
        assert_eq!(data, [0xa5; 4]);
        assert!(bus.read(u64::MAX, &mut data), "the saturated top address must dispatch");
        assert!(!bus.read(u64::MAX - 16, &mut data), "one below the base must not");
    }

    #[test]
    fn find_device_returns_the_same_handle_offset_and_irq_as_a_direct_dispatch() {
        let mut bus = Bus::new();
        let dev = dummy();
        bus.register(0x2000, 0x100, dev.clone(), Some(7));

        let (found, offset, irq) = bus.find_device(0x2010).expect("must find the registered device");
        assert!(Arc::ptr_eq(&found, &dev), "must be the same underlying device, not a copy");
        assert_eq!(offset, 0x10);
        assert_eq!(irq, Some(7));

        assert!(bus.find_device(0x3000).is_none(), "an unmapped address must find nothing");
    }

    #[test]
    fn re_registering_a_device_moves_its_mapping_instead_of_duplicating_it() {
        let mut bus = Bus::new();
        let dev = dummy();
        bus.register(0x1000, 0x100, dev.clone(), None);
        bus.register(0x2000, 0x100, dev.clone(), None);

        let mut data = [0u8; 1];
        assert!(!bus.read(0x1000, &mut data), "the old mapping must be gone");
        assert!(bus.read(0x2000, &mut data), "the new mapping must be live");
    }

    #[test]
    fn unregister_removes_a_mapping_entirely() {
        let mut bus = Bus::new();
        let dev = dummy();
        bus.register(0x1000, 0x100, dev.clone(), Some(9));
        assert_eq!(bus.write(0x1000, &[1]), Some(Some(9)));

        bus.unregister(&dev);
        assert_eq!(bus.write(0x1000, &[1]), None, "nothing should answer at a reclaimed address");
    }

    #[test]
    fn an_overlapping_registration_is_refused_and_leaves_the_original_working() {
        let mut bus = Bus::new();
        bus.register(0x1000, 0x100, dummy(), None);
        bus.register(0x1080, 0x100, dummy(), None); // overlaps the tail of the first

        let mut data = [0u8; 1];
        assert!(bus.read(0x10ff, &mut data), "the original mapping must survive");
        assert!(!bus.read(0x1100, &mut data), "the refused mapping must not be live");
    }
}
